urllib3/response.py

```diff
@@ -63,7 +63,12 @@ def __getattr__(self, name):
     def decompress(self, data):
         if not data:
             return data
-        return self._obj.decompress(data)
+        ret = self._obj.decompress(data)
+        while self._obj.unused_data:
+            (self._obj, data) = (zlib.decompressobj(16 + zlib.MAX_WBITS),
+                                 self._obj.unused_data)
+            ret += self._obj.decompress(data)
```

Comment:
So if there is junk at the end of a gzipped file for any reason, this change will break that. I wonder how much we care about it: should we aim to tolerate errors here?

Reply:
I suggest doing a simple test like this: Construct a file that has some junk at the end of the gzip file, put it to some web server, and see how the popular browsers behave for them.  If they report errors for these files, we can just ignore the issue and let it raise an exception.  In this case we are fixing the problem in these cases too: we should have reported an error but we haven't.

Reply:
@isaacto great. Can you perform that test and report back?

Reply:
I tried firefox, which doesn't even handle a multi-member gzip-encoded HTML file.  I found that there is a bug report about a similar issue in Chrome, which didn't seem to have been resolved (see https://bugs.chromium.org/p/chromium/issues/detail?id=20884).

I tried the gzip program, and found that it can report trailing garbage in case of errors:

```
$ echo hello > f
$ gzip f
$ echo foo >> f.gz
$ gunzip f.gz

gzip: f.gz: decompression OK, trailing garbage ignored
```

That made me wonder a bit about whether we should add proper handling of gzip file in urllib3.  But since, unlike a regular browser, urllib3 automatically decompress downloaded files in its usual usage, I think the files should be properly handled.  (I discovered this problem when I tried downloading large files from Reuters, so we can expect that other users will sooner or later discover the same issue.)

If we really want to add trailing garbage tolerance to the code, we need quite a bit of machinery.  In particular, we need (1) to check whether an error from zlib is due to missing header, and (2) to check whether there is some content corrected decompressed.  If both answers are "yes", then we can tolerate the error.  To do so, we need to mark the decoder with a state which will ignore further content.

I wonder whether this complexity is desirable.

Reply:
> But since, unlike a regular browser, urllib3 automatically decompress downloaded files in its usual usage

Well, this isn't quite true. urllib3 only decompresses files that are marked as having compression that is not part of the core representation (that is, with `Content-Encoding: gzip`), which should not be applied to files whose core representation is a compressed file.

Nonetheless, a few web servers do like doing that, so it can happen.

Given that it seems that browsers don't support multi-member gzip I think merely supporting that use-case is sufficient. I'm ok swallowing errors on garbage trailing data. It should be easy enough to do that: a single boolean flag is probably sufficient.

Reply:
See whether my new PR looks better.